src/dscanner/analysis/static_if_else.d

```diff
@@ -22,43 +19,50 @@ import dscanner.utils : safeAccess;
  * 
  * However, it's more likely that this is a mistake.
  */
-final class StaticIfElse : BaseAnalyzer
+extern(C++) class StaticIfElse(AST) : BaseAnalyzerDmd!AST
 {
-	alias visit = BaseAnalyzer.visit;
-
+	alias visit = BaseAnalyzerDmd!AST.visit;
 	mixin AnalyzerInfo!"static_if_else_check";
 
-	this(string fileName, bool skipTests = false)
+	extern(D) this(string fileName, bool skipTests = false)
 	{
-		super(fileName, null, skipTests);
+		super(fileName, skipTests);
 	}
 
-	override void visit(const ConditionalStatement cc)
+	override void visit(AST.ConditionalStatement s)
 	{
-		cc.accept(this);
-		if (cc.falseStatement is null)
-		{
-			return;
-		}
-		const(IfStatement) ifStmt = getIfStatement(cc);
-		if (!ifStmt)
+		import dmd.astenums : STMT;
+		staticIfEncountered = false;
+
+		s.condition.accept(this);
+		if (s.ifbody)
+            s.ifbody.accept(this);
+        
+		if (s.elsebody)
 		{
-			return;
+			if (staticIfEncountered && s.elsebody.stmt == STMT.If)
+				addErrorMessage(cast(ulong) s.elsebody.loc.linnum, cast(ulong) s.elsebody.loc.charnum,
+					KEY, MESSAGE);
+		
+			s.elsebody.accept(this);
 		}
-		addErrorMessage(ifStmt.line, ifStmt.column, KEY, "Mismatched static if. Use 'else static if' here.");
 	}
 
-	const(IfStatement) getIfStatement(const ConditionalStatement cc)
+	override void visit(AST.StaticIfCondition c)
 	{
-		return safeAccess(cc).falseStatement.statement.statementNoCaseNoDefault.ifStatement;
+		staticIfEncountered = true;
+		super.visit(c);
```

Comment:
I don't understand the logic here. Why are we visiting the super in `StaticIfCondition` ? This doesn't seem to do anything.

I know that the compiler is a bit contrived in dealing with `static if`. However, by looking at the parser we can clarify things: it seems that the parser creates `StaticIfDeclaration` if the `static if` is encountered outside the scope of a function and `ConditionalStatement` if the `static if` is encountered inside a function scope. Both AST nodes contain information about the the else body so you can check there if the else body has exactly one statement and that statement is an `if`.

Reply:
Sorry for that, I think for this visitor I should have added a more thorough explanation:
Firstly, regarding `StaticIfDeclaration`, I did not override it because inside the global scope, the construction mentioned in the example makes no sense (while we can have `static if` at the global scope, we can't have simple `if`), that's why I omitted that.
Secondly, I checked `StaticIfCondition` because the `ConditionalStatement` has the `condition` field, which can be a `StaticIfCondition` or something else, like `version` for example. I would agree this is not the most straightforward and elegant approach, but currently I don't see other approach.
Do you think it would make sense to have a method `isStaticIfCondition`, (there are plenty of similar methods, like `isClassDeclaration`, but in `Dsymbol` class, not `Condition`), or is that too specific for our use case?